src/topology/uniform_space/separation.lean

```diff
@@ -51,23 +113,65 @@ theorem separated_def' {α : Type u} [uniform_space α] :
 separated_def.trans $ forall_congr $ λ x, forall_congr $ λ y,
 by rw ← not_imp_not; simp [classical.not_forall]
 
+lemma id_rel_sub_separation_relation (α : Type*) [uniform_space α]: id_rel ⊆ 𝓢 α :=
+begin
+  unfold separation_rel,
+  rw id_rel_subset,
+  intros x,
+  suffices : ∀ t ∈ 𝓤 α, (x, x) ∈ t, by simpa only [refl_mem_uniformity],
+  exact λ t, refl_mem_uniformity,
+end
+
+lemma separation_rel_comap  {f : α → β} (h : ‹uniform_space α› = uniform_space.comap f ‹uniform_space β›) :
+𝓢 α = (prod.map f f) ⁻¹' 𝓢 β :=
+begin
+  dsimp [separation_rel],
+  rw [uniformity_comap h, (filter.comap_has_basis (prod.map f f) (𝓤 β)).sInter_sets,
+      ← preimage_bInter, sInter_eq_bInter],
+  refl,
+end
+
+protected lemma filter.has_basis.separation_rel {ι : Type*} {p : ι → Prop} {s : ι → set (α × α)}
+  (h : has_basis (𝓤 α) p s) :
+𝓢 α = ⋂ i ∈ set_of p, s i :=
+by { unfold separation_rel, rw h.sInter_sets }
+
+lemma separation_rel_eq_inter_closure : 𝓢 α = ⋂₀ (closure '' (𝓤 α).sets) :=
+by simpa [uniformity_has_basis_closure.separation_rel]
+
+lemma is_closed_separation_rel : is_closed (𝓢 α) :=
+begin
+  rw separation_rel_eq_inter_closure,
+  apply is_closed_sInter,
+  rintros _ ⟨t, t_in, rfl⟩,
+  exact is_closed_closure,
+end
+
 @[priority 100] -- see Note [lower instance priority]
 instance separated_t2 [s : separated α] : t2_space α :=
-⟨assume x y, assume h : x ≠ y,
-let ⟨d, hd, (hxy : (x, y) ∉ d)⟩ := separated_def'.1 s x y h in
-let ⟨d', hd', (hd'd' : comp_rel d' d' ⊆ d)⟩ := comp_mem_uniformity_sets hd in
-have {y | (x, y) ∈ d'} ∈ 𝓝 x,
-  from mem_nhds_left x hd',
-let ⟨u, hu₁, hu₂, hu₃⟩ := mem_nhds_sets_iff.mp this in
-have {x | (x, y) ∈ d'} ∈ 𝓝 y,
-  from mem_nhds_right y hd',
-let ⟨v, hv₁, hv₂, hv₃⟩ := mem_nhds_sets_iff.mp this in
-have u ∩ v = ∅, from
-  eq_empty_of_subset_empty $
-  assume z ⟨(h₁ : z ∈ u), (h₂ : z ∈ v)⟩,
-  have (x, y) ∈ comp_rel d' d', from ⟨z, hu₁ h₁, hv₁ h₂⟩,
-  hxy $ hd'd' this,
-⟨u, v, hu₂, hv₂, hu₃, hv₃, this⟩⟩
+begin
+  rw [t2_iff_is_closed_diagonal, ← show 𝓢 α = diagonal α, from s],
+  exact is_closed_separation_rel
+end
+
+lemma separated_iff_t2 : separated α ↔ t2_space α :=
```

Comment:
At first reading, I was wondering why you would need to have two different typeclasses if they are equal (I mean, having it as a definition would be fine, but the statements could be reformulated with the typeclass assumption `[t2_space]` instead). However, I realized that this would create an instance loop `t3_space -> t2_space` and `t2_space + uniform_space -> t3_space`. If this is one of the reasons for this design, it would be worth mentioning somewhere (for instance in a docstring to this lemma).

Reply:
I have no idea about the reasons for this design, we need to ask @johoelzl.